kubespawner/objects.py

```diff
@@ -829,31 +831,43 @@ def make_ingress(
         labels=ingress_labels,
     )
 
+    ingress_specifications = ingress_specifications or [{}]
+
+    rules = [
+        V1IngressRule(
+            host=host or spec.get("host"),
```

Comment:
Ah now I see this involved a reference to `host`.

Should this be `spec.get("host") or host` instead though? I'm not sure at all, just wanted to make sure that at least one of us is sure about what makes sense.

Reply:
Yeah I think this is currently weird, considering how `spec["host"]` is referenced in the tls section.

Reply:
Since this only affects the KubeIngressProxy part etc, I figure we go with what you think makes sense here - but maybe you could add a comment about this as this is apparently quite tricky?

Reply:
Overall, this PR LGTM, but this makes me uncertain.

I don't overview when host is None, or if there are implicit assumptions about what ingress_specification can and can't be, such as setting just `tlsSecret` and not the `host` etc, and if passing `host=None` is acceptable or not here.

If a comment clarifies it, even though its a comment indicating that we are not sure, that is a clear improvement on managing this complexity I think.

Reply:
IMHO, the best way here is to merge host from `routespec` with items from `ingress_specifications`. But I'm still not sure about the way of merging these values.

For example, `routespec=https://my.example.com/me/service` and `ingress_specifications=[{"host": "*.example.com", "tlsSecretName": "some-secret"}]`. Do we need to create:
```yaml
spec:
  rules:
    - host: my.example.com
      path: /me/service
    - host: *.example.com
      path: /me/service
  tls:
    hosts:
      - *.example.com
    secretName: my-secret
```
or
```yaml
spec:
  rules:
    - host: *.example.com
      path: /me/service
  tls:
    hosts:
      - *.example.com
    secretName: my-secret
```
or
```yaml
spec:
  rules:
    - host: my.example.com
      path: /me/service
  tls:
    hosts:
      - *.example.com
    secretName: my-secret
```

Reply:
Options 1 and 2 can lead to the following issue.

Example:
```python
c.Jupyterhub.subdomain_host="example.com"
c.KubeIngressProxy.ingress_specifications = [{"host": "*.example.com", "tlsSecretName": "secret-name"}]
```
For user1 and user2 Jupyterhub will create `routespec="https://user1.example.com/default-server"` and routespec="https://user2.example.com/default-server"` respectively.

But then both ingress rules will look like:
```
spec:
  rules:
  - host: *.example.com
    path: /default-server
  tls:
    hosts:
    - *.example.com
    secretName: secret-name
```
and they cannot be distinct from each other.

So I've implemented option 3 which produces rules like:
```
spec:
  rules:
  - host: user1.example.com
    path: /default-server
  tls:
    hosts:
    - *.example.com
    secretName: secret-name
```
and added some tests covering these cases.